Grid/qcd/utils/GaugeGroupTwoIndex.h

```diff
@@ -108,60 +178,53 @@ class GaugeGroupTwoIndex : public GaugeGroup<ncolour, group_name> {
     }
 
     if (Index < ncolour * (ncolour - 1) / 2) {
-      baseOffDiagonal(a[Index][0], a[Index][1], eij, group_name());
+      baseOffDiagonal(a[Index][0], a[Index][1], eij);
     } else {
       baseDiagonal(Index, eij);
     }
   }
+    
+  template <class cplx>
+   static void base(int Index, iGroupMatrix<cplx> &eij, GroupName::Sp) {
+     // returns (e)^(ij)_{kl} necessary for change of base U_F -> U_R
+     assert(Index < Dimension);
+     eij = Zero();
+
+     // for the linearisation of the 2 indexes
+     static int a[ncolour * (ncolour - 1) / 2][2];  // store the a <-> i,j
+     static bool filled = false;
+     if (!filled) {
+       int counter = 0;
+       for (int i = 1; i < ncolour; i++) {
+         for (int j = 0; j < i; j++) {
+           a[counter][0] = i;
+             if (j==0 && i==ngroup+j && S==-1) {
+                 //std::cout << "skipping" << std::endl; // for Sp2n this vanishes identically.
+                 j = j+1;
+             }
+           a[counter][1] = j;
+           counter++;
+         }
+       }
+       filled = true;
+     }
+
+     if (Index < DimensionHelper<ncolour, AntiSymmetric, GroupName::Sp>::Dimension + 1) {   // +1 the singlet
+       baseOffDiagonalSp(a[Index][0], a[Index][1], eij);
+     } else {
+       baseDiagonal(Index, eij);
+     }
+   }
 
   template <class cplx>
   static void baseDiagonal(int Index, iGroupMatrix<cplx> &eij) {
     eij = Zero();
     eij()()(Index - ncolour * (ncolour - 1) / 2,
             Index - ncolour * (ncolour - 1) / 2) = 1.0;
   }
-
+    
   template <class cplx>
-  static void baseOffDiagonal(int i, int j, iGroupMatrix<cplx> &eij,
-                              GroupName::Sp) {
-    eij = Zero();
-    RealD tmp;
-
-    if ((i == ngroup + j) && (1 <= j) && (j < ngroup)) {
-      for (int k = 0; k < ngroup; k++) {
-        if (k < j) {
-          tmp = sqrt(2 * j * (j + 1));
-          tmp = 1 / tmp;
-          tmp *= std::sqrt(2.0);
-          eij()()(k, k + ngroup) = tmp;
-          eij()()(k + ngroup, k) = -tmp;
-        }
-        if (k == j) {
-          tmp = sqrt(2 * j * (j + 1));
-          tmp = -j / tmp;
-          tmp *= std::sqrt(2.0);
-          eij()()(k, k + ngroup) = tmp;
-          eij()()(k + ngroup, k) = -tmp;
-        }
-      }
-
-    }
-
-    else if (i != ngroup + j) {
-      for (int k = 0; k < ncolour; k++)
-        for (int l = 0; l < ncolour; l++) {
-          eij()()(l, k) =
-              delta(i, k) * delta(j, l) + S * delta(j, k) * delta(i, l);
-        }
-    }
-
-    RealD nrm = 1. / std::sqrt(2.0);
-    eij = eij * nrm;
-  }
-
-  template <class cplx>
-  static void baseOffDiagonal(int i, int j, iGroupMatrix<cplx> &eij,
-                              GroupName::SU) {
+  static void baseOffDiagonal(int i, int j, iGroupMatrix<cplx> &eij) {
```

Comment:
Why is this distinguished from its Sp counterpart by name and not by function signature?

Reply:
if it can be done we should do it!

Reply:
What was hindering `baseOffDiagonal(..., group_name())` strategy as everywhere else?

Reply:
Okay. But in this case you are still only specialising in the group at first and then you specialise in the symmetry via reference to your helpers. I'm only talking about the first step here.